### PR TITLE
Change default TLS behavior in dispatch to favor is-a style

### PR DESCRIPTION
**Pull request by albanD**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #75827

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
Thanks!

**Comment by albanD** [comment]
FYI @zou3519 @bdhirsh for changes to composite compliance and functionalization tests

**Comment by albanD** [comment]
Note that there is a functorch side PR as well: https://github.com/pytorch/functorch/pull/718 that will be landed alongside this one.

File: aten/src/ATen/core/PythonFallbackKernel.cpp [hunk 1]
```diff
@@ -58,8 +45,10 @@ struct StashTLSOnEntryGuard {
 
 void pythonFallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
   TORCH_INTERNAL_ASSERT(tls_on_entry.has_value());
-  c10::impl::ForceDispatchKeyGuard dispatcher_guard(tls_on_entry.value());
-  StashTLSOnEntryGuard stash_guard;
+  // c10::impl::ForceDispatchKeyGuard dispatcher_guard(tls_on_entry.value());
+  // StashTLSOnEntryGuard stash_guard;
```
Comment (zou3519) [inline aten/src/ATen/core/PythonFallbackKernel.cpp:49]: @albanD why is this commented out? Can we delete it?

Comment (albanD) [reply aten/src/ATen/core/PythonFallbackKernel.cpp:49]: Yes it can be removed. The restore that the user has access to is doing this (if they ask for it).

**Pull request closed without merging**